core/trino-spi/src/main/java/io/trino/spi/type/SqlTimeWithTimeZone.java

```diff
@@ -103,20 +103,38 @@ public int hashCode()
     @Override
     public String toString()
     {
-        StringBuilder builder = new StringBuilder();
-        builder.append(format(
-                "%02d:%02d:%02d",
-                picos / PICOSECONDS_PER_HOUR,
-                (picos / PICOSECONDS_PER_MINUTE) % MINUTES_PER_HOUR,
-                (picos / PICOSECONDS_PER_SECOND) % SECONDS_PER_MINUTE));
+        StringBuilder builder = new StringBuilder(14 + (precision == 0 ? 0 : 1 + precision));
+        appendTwoDigits((int) (picos / PICOSECONDS_PER_HOUR), builder);
+        builder.append(':');
+        appendTwoDigits((int) ((picos / PICOSECONDS_PER_MINUTE) % MINUTES_PER_HOUR), builder);
+        builder.append(':');
+        appendTwoDigits((int) ((picos / PICOSECONDS_PER_SECOND) % SECONDS_PER_MINUTE), builder);
 
         if (precision > 0) {
             long scaledFraction = (picos % PICOSECONDS_PER_SECOND) / POWERS_OF_TEN[MAX_PRECISION - precision];
-            builder.append(".");
-            builder.append(format("%0" + precision + "d", scaledFraction));
+            builder.append('.');
+            builder.setLength(builder.length() + precision);
+
+            for (int index = builder.length() - 1; index > 8; index--) {
+                long temp = scaledFraction / 10;
+                int digit = (int) (scaledFraction - (temp * 10));
+                scaledFraction = temp;
+                builder.setCharAt(index, (char) ('0' + digit));
```

Comment:
Didn't we add similar code somewhere else recently? We may want to consider putting it somewhere we can reuse it, since it's not particularly trivial.

Reply:
We did indeed. That change was to SqlTimestamp in the SPI which improved performance just for the JSON serialization path. I think there are a few problems that I think has caused this code to proliferate:
- SqlTimestamp and SqlTimestampWithTimeZone are SPI objects that seem to only exist in order to serialize values as JSON. They have similar, but slightly different logic from one another (the time zone part)
- Casting to timestamp (and with time zone) types are currently implemented in trino-main and also formatted to strings (even though casting to varchar needs a UTF8 slice) because the formatting logic is being shared for long and short timestamp types both with and without time zones and for conversions and in parts of the JSON type handling paths. So it’s a 2x2 grid of timestamp types each targeting at least 2 output representations, and not accessible from the SPI - meanwhile the SPI can’t host the implementation because it depends on the packing / unpacking implementations in main.
- Time with and without time zone is similar in Trino main, except that was already producing a slice directly (presumably for performance to avoid the intermediate string and since the format is simple enough to implement by hand) which means it also can’t live in the SPI because of the dependency on slice.